corehq/apps/case_importer/do_import.py

```diff
@@ -0,0 +1,278 @@
+            raise exceptions.BlankExternalId()
+
+    def relies_on_uncreated_case(self, uncreated_external_ids):
+        return any(lookup_id and lookup_id in uncreated_external_ids
+                   for lookup_id in [self.search_id, self.parent_id, self.parent_external_id])
+
+    def set_owner_id(self, name_cache, id_cache):
+        owner_id = self.uploaded_owner_id
+        if self.uploaded_owner_name:
+            # If an owner name was provided, use the id of the provided or
+            # owner rather than the uploaded_owner_id
+            try:
+                owner_id = importer_util.get_id_from_name(
+                    self.uploaded_owner_name, self.domain, name_cache
+                )
+            except SQLLocation.MultipleObjectsReturned:
+                raise exceptions.DuplicateLocationName()
+
+            if not owner_id:
+                raise exceptions.InvalidOwnerName('owner_name')
+
+        if owner_id:
+            # If an owner_id mapping exists, verify it is a valid user
+            # or case sharing group
+            if importer_util.is_valid_id(owner_id, self.domain, id_cache):
+                id_cache[owner_id] = True
+            else:
+                id_cache[owner_id] = False
+                raise exceptions.InvalidOwnerId('owner_id')
+
+        # if they didn't supply an owner, default to current user
+        self.owner_id = owner_id or self.user_id
+
+    def set_parent_id(self, log_case_lookup):
+        for column, search_field, search_id in [
+                ('parent_id', 'case_id', self.parent_id),
+                ('parent_external_id', 'external_id', self.parent_external_id),
+        ]:
+            if search_id:
+                parent_case, error = importer_util.lookup_case(
+                    search_field, search_id, self.domain, self.parent_type)
+                log_case_lookup()
+                if parent_case:
+                    self.extras['index'] = {
+                        self.parent_ref: (parent_case.type, parent_case.case_id)
+                    }
+                    return
+                else:
+                    raise exceptions.InvalidParentId(column)
+
+    def set_date_opened(self):
+        if self.date_opened and BULK_UPLOAD_DATE_OPENED.enabled(self.domain):
+            self.extras['date_opened'] = self.date_opened
+
+    def set_external_id(self, is_new_case):
+        if is_new_case and self.config.search_field == 'external_id':
+            self.extras['external_id'] = self.search_id
+        elif self.external_id:
+            self.extras['external_id'] = self.external_id
+
+    def get_create_caseblock(self):
+        return CaseBlock(
+            create=True,
+            case_id=uuid.uuid4().hex,
+            owner_id=self.owner_id,
+            user_id=self.user_id,
+            case_type=self.config.case_type,
+            case_name=self.case_name or '',
+            update=self.fields_to_update,
+            **self.extras
```

Comment:
building on my comment from before - I think there might be a way that you can just generate these when you call `self.extras`, so it would be `self.get_extras()` which would then return all of the appropriate items so that the `_Importer` class doesn't have to populate them.